compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -550,30 +550,21 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
             } else {
                 format!("{}.dll", name)
             };
-            dylib_table
-                .entry(name)
-                .or_default()
-                .extend(lib.dll_imports.iter().map(|import| import.name));
+            dylib_table.entry(name).or_default().extend(lib.dll_imports.iter().cloned());
         }
     }
 
-    // FIXME: when we add support for ordinals, fix this to propagate ordinals.  Also figure out
-    // what we should do if we have two DllImport values with the same name but different
-    // ordinals.
+    // Rustc already signals an error if we have two imports with the same name but different
+    // calling conventions (or function signatures), so we don't have pay attention to those
+    // when ordering.
+    // FIXME: when we add support for ordinals, figure out if we need to do anything if we
+    // have two DllImport values with the same name but different ordinals.
     let mut result = dylib_table
         .into_iter()
-        .map(|(lib_name, imported_names)| {
-            let mut names = imported_names
-                .iter()
-                .map(|name| DllImport { name: *name, ordinal: None })
-                .collect::<Vec<_>>();
-            names.sort_unstable_by(|a: &DllImport, b: &DllImport| {
-                match a.name.as_str().cmp(&b.name.as_str()) {
-                    Ordering::Equal => a.ordinal.cmp(&b.ordinal),
-                    x => x,
-                }
-            });
-            (lib_name, names)
+        .map(|(lib_name, import_set)| {
+            let mut imports = Vec::from_iter(import_set.into_iter());
+            imports.sort_unstable_by_key(|x: &DllImport| x.name.as_str());
+            (lib_name, imports)
         })
         .collect::<Vec<_>>();
     result.sort_unstable_by(|a: &(String, Vec<DllImport>), b: &(String, Vec<DllImport>)| {
```

Comment:
```suggestion
    result.sort_unstable_by_key(...
```

Reply:
Yeah, I've tried to get that to work, but I can't make the lifetimes work (unless I copy the strings, but that seems wasteful).

```
573 |     result.sort_unstable_by_key(|x: &(String, Vec<DllImport>)| &x.0);
    |                                     -                        - ^^^^ returning this value requires that `'1` must outlive `'2`
    |                                     |                        |
    |                                     |                        return type of closure is &'2 std::string::String
    |                                     let's call the lifetime of this reference `'1`
```

Lifting the closure to a named function with lifetime parameters doesn't work either:
```
    fn extract_key<'a, 'b: 'a>(x: &'b (String, Vec<DllImport>)) -> &'a str { &x.0 }

    result.sort_unstable_by_key(extract_key);
```
results in an error saying only "higher-ranked subtype error" at the call site.

Reply:
Ok.